esmvaltool/diag_scripts/weighting/climwip.py

```diff
@@ -0,0 +1,482 @@
+    plt.close(figure)
+
+    filename_data = get_diagnostic_filename(f'independence_{variable}',
+                                            cfg,
+                                            extension='nc')
+    independence.to_netcdf(filename_data)
+
+    caption = f'Euclidean distance matrix for variable {variable}'
+    log_provenance(caption, filename_plot, cfg, ancestors)
+    log_provenance(caption, filename_data, cfg, ancestors)
+
+
+def calculate_performance(model_data: 'xr.DataArray',
+                          obs_data: 'xr.DataArray') -> 'xr.DataArray':
+    """Calculate performance.
+
+    Calculate the area weighted mean between the given ensemble of model
+    data, and observation data. The observation data must have the
+    ensemble dimension squeezed or reduced. Returns an xarray.DataArray
+    containing the same number of values as members of `model_data`.
+    """
+    diff = model_data - obs_data
+
+    performance = area_weighted_mean(diff**2)**0.5
+
+    performance.name = f'd{model_data.name}'
+    performance.attrs['variable_group'] = model_data.name
+    performance.attrs["units"] = model_data.units
+
+    return performance
+
+
+def barplot(metric: 'xr.DataArray', label: str, filename: str):
+    """Visualize metric as barplot."""
+    name = metric.name
+    variable_group = metric.variable_group
+    units = metric.units
+
+    metric_df = metric.to_dataframe().reset_index()
+
+    ylabel = f'{label} {variable_group} ({units})'
+
+    figure, axes = plt.subplots(figsize=(15, 10))
+    chart = sns.barplot(x='model_ensemble', y=name, data=metric_df, ax=axes)
+    chart.set_xticklabels(chart.get_xticklabels(),
+                          rotation=45,
+                          horizontalalignment='right')
+    chart.set_title(f'{label} for {variable_group}')
+    chart.set_ylabel(ylabel)
+    chart.set_xlabel('')
+
+    figure.savefig(filename, dpi=300, bbox_inches='tight')
+    plt.close(figure)
+
+
+def visualize_and_save_performance(performance: 'xr.DataArray', cfg: dict,
+                                   ancestors: list):
+    """Visualize performance."""
+    label = 'RMS error'
+
+    variable_group = performance.variable_group
+    filename_plot = get_plot_filename(f'performance_{variable_group}', cfg)
+
+    barplot(performance, label, filename_plot)
+
+    filename_data = get_diagnostic_filename(f'performance_{variable_group}',
+                                            cfg,
+                                            extension='nc')
+    performance.to_netcdf(filename_data)
+
+    caption = f'Performance metric {label} for variable group {variable_group}'
+    log_provenance(caption, filename_plot, cfg, ancestors)
+    log_provenance(caption, filename_data, cfg, ancestors)
+
+
+def compute_overall_mean(dataset):
+    """Normalize all variables in a dataset and return their mean."""
+    normalized = dataset / dataset.median(dim='model_ensemble')
```

Comment:
For the independence this line leads to the distance matrix becoming asymmetric. For the normalization to work as expected we need to average over both matrix dimensions. Something like this should do the trick:
```
if 'perfect_model_ensemble' in dataset.dims:
    median_dim = ['perfect_model_ensemble', 'model_ensemble']
else:
    median_dim = 'model_ensemble'
normalized = dataset / dataset.median(dim=median_dim)
```